crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -1,280 +1,284 @@
+        match container.kill(signal, true) {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                if container.status() == ContainerStatus::Stopped {
+                    return Err(Error::Others("container not running".into()));
+                }
+                Err(Error::Others(e.to_string()))
+            }
+        }
     }
 
     fn delete(&self) -> Result<(), Error> {
-        let spec = match load_spec(self.bundle.clone()) {
-            Ok(spec) => spec,
+        log::info!("deleting instance: {}", self.id);
+        match instance_exists(&self.rootdir, self.id.as_str()) {
+            Ok(exists) => {
+                if !exists {
+                    return Ok(());
+                }
+            }
             Err(err) => {
-                error!("Could not load spec, skipping cgroup cleanup: {}", err);
+                error!("could not find the container, skipping cleanup: {}", err);
                 return Ok(());
             }
-        };
-        let cg = oci::get_cgroup(&spec)?;
-        cg.delete()?;
+        }
+        let container_root = get_instance_root(&self.rootdir, self.id.as_str())?;
+        let container = Container::load(container_root).with_context(|| {
+            format!(
+                "could not load state for container {id}",
+                id = self.id.as_str()
+            )
+        });
+        match container {
+            Ok(mut container) => container.delete(true).map_err(|err| {
+                Error::Any(anyhow::anyhow!(
+                    "failed to delete container {}: {}",
+                    self.id,
+                    err
+                ))
+            })?,
+            Err(err) => {
+                error!("could not find the container, skipping cleanup: {}", err);
+                return Ok(());
+            }
+        }
+
         Ok(())
     }
 
     fn wait(&self, waiter: &Wait) -> Result<(), Error> {
+        log::info!("waiting for instance: {}", self.id);
         let code = self.exit_code.clone();
         waiter.set_up_exit_code_wait(code)
     }
 }
 
+impl Wasi {
+    fn build_container(
+        &self,
+        stdin: &str,
+        stdout: &str,
+        stderr: &str,
+        engine: Engine,
+    ) -> anyhow::Result<Container> {
+        let syscall = create_syscall();
+        let stdin = maybe_open_stdio(stdin).context("could not open stdin")?;
+        let stdout = maybe_open_stdio(stdout).context("could not open stdout")?;
+        let stderr = maybe_open_stdio(stderr).context("could not open stderr")?;
+
+        let container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
+            .with_executor(vec![Box::new(WasmtimeExecutor {
+                stdin,
+                stdout,
+                stderr,
+                engine,
+            })])?
+            .with_root_path(self.rootdir.clone())?
+            .as_init(&self.bundle)
+            .with_systemd(false)
+            .build()?;
+        Ok(container)
```

Comment:
Not entirely sure the return type of `build()`, but will this work?

```suggestion
        ContainerBuilder::new(self.id.clone(), syscall.as_ref())
            .with_executor(vec![Box::new(WasmtimeExecutor {
                stdin,
                stdout,
                stderr,
                engine,
            })])?
            .with_root_path(self.rootdir.clone())?
            .as_init(&self.bundle)
            .with_systemd(false)
            .build()
```

Reply:
Unfortunately this does not work as the `build` returns a `LibcontainerError`